docs/adr/0007-seperate-binding-generators.md

```diff
@@ -0,0 +1,142 @@
+* Bad, because it will be hard to guarantee the quality of binding generators since they won't depend on any traits from `uniffi_bindgen`
+* Bad, because it complicates version compatibility, and we will need more assertions to guarantee it
+* Bad, because testability for might increase in complexity. Bindings not in this repository will not have access to the extensive fixtures and examples we have. We might need to expose those if take this option.
+* Bad because it increases the number of crates we manage. As we would manage the core crates for the bindings we depend on (i.e kotlin, swift, python, and soon a C++/gecko one). This will make the release process painful, but we can add automation and clear crate ownership to help.
+
+
+
+## Decision Outcome
+
+Chosen option:
+
+*  **[Option 2] Expose a trait in `uniffi_bindgen` that can be implemented by other crates, to separate the implementation of the bindings**
+
+### Changes Proposed
+#### The exposed trait for binding generators to implement
+`unffi_bindgen` would define a public trait, `BindingGenerator` that includes one main function:
+```rs
+/// Takes in the ComponentInterface and writes the bindings into directory `out_dir`
+fn write_bindings<P: AsRef<Path>>(&self, ci: &CompenentInterface, out_dir: P) -> Result<()>;
+```
+
+Additionally, we can define optional functions to extend testing functionality
+```rs
+/// Compiles the generated bindings using `write_bindings`
+fn compile_bindings<P: AsRef<Path>>(&self, ci: &ComponentInterface, out_dir: P) -> Result<()>;
+
+/// Runs a script in the language that implements the `BindingGenerator` trait
+fn run_script<P: AsRef<Path>>(&self, out_dir: P, script_file: P) -> Result<()>;
+```
+
+There are a few specific ways to implement this
+
+## **Implementation Option 1: Have `uniffi_bindgen` shell out to `uniffi_bindgen_{language}`**
+Using this approach, we would recommend users to install both the `uniffi_bindgen` **binary** and the `uniffi_bindgen_kotlin` (using kotlin as an example) **binary**. Then, the user would run the following:
+1. The user runs `uniffi-bindgen generate -l kotlin -o <OUT_PATH> <UDL_PATH>`
+1. `uniffi-bindgen` gets its own version, let's say `0.16` as an example.
+1. `uniffi-bingen` shells out to `uniffi-bingen-kotlin -o <OUT_PATH> -v 0.16 <UDL_PATH>`
+1. `uniffi-bindgen-kotlin` asserts that it was compiled using the same `0.16` version, otherwise it panics.
+1. `uniffi-bindgen-kotlin` then calls a `generate_bindings` function in `uniffi-bindgen` (as a library dependency)
+1. `uniffi-bindgen`'s `generate_bindings` takes a generic parameter type that implements a trait `BindingGenerator`.
+1. `uniffi-bindgen` parses the `udl`, then calls the generic type's `write_bindings` function (the implementation lives in `uniffi-bingen-kotlin`)
+
+
+## **Implementation Option 2: Have `uniffi_bindgen` get dynamically linked with exactly one `uniffi_bindgen_{language}`**
```

Comment:
Why "exactly one"?

Reply:
Eeeh I'm a little fuzzy on this whole option but since it's "possible" thought I'd note it down - the idea is `uniffi_bindgen` would call an `extern C` function without knowledge of who actually implements it.. and based on who it got linked with it "just works" - I have little experience and comfort with this type of communication so 😅